File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
var BSTIterator = function(root) {
    
    this.pointer = -1;
    this.nodes = [];
    let inorder = (node) => {
        if(node) {
            inorder(node.left);
             this.nodes.push(node);
            inorder(node.right);
        }
    }
    inorder(root);
};

/**
 * @return {number}
 */
BSTIterator.prototype.next = function() {
    this.pointer++;
    return this.nodes[this.pointer].val;
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function() {
    if(this.pointer + 1 < this.nodes.length) {
        return true;
    }
    return false;
};

/** 
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */

File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {

    public int pointer = -1;
    public ArrayList<TreeNode> nodes;
    public void inorder(TreeNode node) {
        if(node != null) {
            inorder(node.left);
            nodes.add(node);
            inorder(node.right);
        }
    }
    public BSTIterator(TreeNode root) {
        nodes = new ArrayList<TreeNode>();
        inorder(root);
    }
    
    public int next() {
        return nodes.get(++pointer).val;
    }
    
    public boolean hasNext() {
        return (pointer + 1) < nodes.size() ? true : false;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */